### PR TITLE
small performance patch

### PR DESCRIPTION
**Pull request by vanhauser-thc**
simple patch: do not instrument functions that only have one basic block. It just pollutes the map.
if there is a decision to call that function it is in the callee and therefore covered.

### CODE REVIEW

**Comment by pzread** [comment]
Thanks for the patch! Since we don't have any unittest :sweat_smile: Please give me a few days to verify it won't break anything. Looks good to me now.

**Comment by pzread** [comment]
After thinking a while, I think if we don't instrument the single block functions, in the case of indirect call, we will lose the track on the coverage of those target functions(if some of them are single block functions). The target functions of an indirect call are especially important to us, because they usually do something special.

So I prefer to keep instrumenting all functions. What do you think?

**Comment by vanhauser-thc** [comment]
you only track decisions with the instrumentation. there is no decision in a single block function.
if an indirect call calls such a function - what is the difference to if that single block would follow that block instead of being an indirect call? None - and that would not be instrumented either. And if there is a decision to take the indirect call or not that would be in the callee ... so I stand by my point that the instrumentation is useless :)

**Comment by pzread** [comment]
There is a case that touching new single block functions from an indirect call needs to be tracked.
For example:
```c
void single_block_fun1() {
  // some side effects
}
void single_block_fun2() {
  // some side effects
}
void single_block_fun3() {
  // some side effects
}
void (*func_ptr[3])() = {single_block_fun1, single_block_fun2, single_block_fun3};
void foo(int input) {
  if (input >= 0 && input <= 2) {
    func_ptr[input]();
  }
}
```
In this case, the side effects in those single block functions might be interesting to us (although they haven't triggered any other branches, they might be the first step to satisfy more branch conditions, as we actually reached more code).
So if fuzzer generated an input which covers them, we should save the input. If we choose not to instrument those single block functions, the fuzzer won't be able to see the differences between `input=0/1/2`, if we don't do other instrumentation on indirect call (e.g. SanitizerCoverage directly instruments the indirect call for tracking).

**Comment by vanhauser-thc** [comment]
Hi,
that sounds like a valid argument :)
A jump table within a function would not be an issue, as the successors are visible in the CFG.
However a call table is different. 
Thinking more about it - I still think it is unnecessary as in your example `input` has to be set somewhere with the different values, and that would be different paths/basic blocks.
However in some circumstances the value for the variable may come from the input and then this would not be the case. Or it comes from a path that is used often hence the corresponding map location would just have a different count and hence afl-fuzz would ignore the difference.
On the other hand instrumenting all such 1 block functions pollutes the too small map (so creating even more collisions than there already are) and slows down the target execution, and that for a very special corner case.
So basically documenting my thought process here I think the right solution to this is making this configurable like LOOPHEAD is the answer IMHO. For C++ targets it is good to have, for most C targets it isn't ...
What do you think?

**Comment by pzread** [comment]
I think this is a good idea :) You're free to update this patch and add the configurable option, then I'll accept the PR. Otherwise I will find some time to do it this week.

**Comment by vanhauser-thc** [comment]
I could do that next week, I am going on vacation in a few hours :)

**Comment by vanhauser-thc** [comment]
here you go (the afl++ parameter for instrim is AFL_LLVM_INSTRIM_SKIPSINGLEBLOCK so I chose SKIPSINGLEBLOCK)

File: LLVMInsTrim.cpp [hunk 1]
```diff
@@ -31,6 +31,9 @@ namespace {
       return generator() % 65536;
     }
 
+ protected:
+  uint32_t function_minimum_size = 1;
```
Comment (pzread) [inline LLVMInsTrim.cpp]: nit: The indent is not aligned for these two lines.
nit: function_minimum_size -> FunctionMinimumSize (we try to follow the LLVM coding style.)

**Comment by vanhauser-thc** [comment]
oh damn you are right. done.

btw I noticed another issue:
```
unsigned genLabel() {
    return generator() % 65536;
```
If the MAP_SIZE_POW2  in config.h is not the default of 16 this can create problems.

**Comment by pzread** [comment]
LGTM. Yes, the MAP_SIZE_POW2 should be fixed. It needs to change the way to build the pass a little bit. I'll take a look later.
